dll/win32/shell32/droptargets/CFSDropTarget.cpp

```diff
@@ -547,104 +547,124 @@ HRESULT CFSDropTarget::_DoDrop(IDataObject *pDataObject,
+            // We need to create a link for each pidl in the copied items,
+            // so step through the pidls from the clipboard.
             for (UINT i = 0; i < lpcida->cidl; i++)
             {
-                //Find out which file we're copying
+                // Find out which file we're linking.
                 STRRET strFile;
                 hr = psfFrom->GetDisplayNameOf(apidl[i], SHGDN_FORPARSING, &strFile);
-                if (FAILED(hr))
-                {
-                    ERR("Error source obtaining path");
+                if (FAILED_UNEXPECTEDLY(hr))
                     break;
-                }
 
-                hr = StrRetToBufW(&strFile, apidl[i], wszPath, _countof(wszPath));
-                if (FAILED(hr))
-                {
-                    ERR("Error putting source path into buffer");
+                hr = StrRetToBufW(&strFile, apidl[i], wszSrcPath, _countof(wszSrcPath));
+                if (FAILED_UNEXPECTEDLY(hr))
                     break;
-                }
-                TRACE("source path = %s", debugstr_w(wszPath));
-
-                // Creating a buffer to hold the combined path
-                WCHAR buffer_1[MAX_PATH] = L"";
-                WCHAR *lpStr1;
-                lpStr1 = buffer_1;
 
-                LPWSTR pwszFileName = PathFindFileNameW(wszPath);
-                LPWSTR pwszExt = PathFindExtensionW(wszPath);
-                LPWSTR placementPath = PathCombineW(lpStr1, m_sPathTarget, pwszFileName);
-                CComPtr<IPersistFile> ppf;
+                TRACE("source path: %s\n", debugstr_w(wszSrcPath));
 
-                //Check to see if it's already a link.
-                if (!wcsicmp(pwszExt, L".lnk"))
+                WCHAR wszDisplayName[MAX_PATH];
+                LPWSTR pwszFileName = PathFindFileNameW(wszSrcPath);
+                if (PathIsRootW(wszSrcPath)) // Drive?
                 {
-                    //It's a link so, we create a new one which copies the old.
-                    if(!_GetUniqueFileName(placementPath, pwszExt, wszTarget, TRUE))
-                    {
-                        ERR("Error getting unique file name");
-                        hr = E_FAIL;
+                    hr = psfFrom->GetDisplayNameOf(apidl[i], 0, &strFile);
+                    if (FAILED_UNEXPECTEDLY(hr))
                         break;
-                    }
-                    hr = IShellLink_ConstructFromPath(wszPath, IID_PPV_ARG(IPersistFile, &ppf));
-                    if (FAILED(hr)) {
-                        ERR("Error constructing link from file");
+
+                    hr = StrRetToBufW(&strFile, apidl[i], wszDisplayName, _countof(wszDisplayName));
+                    if (FAILED_UNEXPECTEDLY(hr))
                         break;
+
+                    // Delete ':' in wszDisplayName.
+                    LPWSTR pch0, pch1;
+                    for (pch0 = pch1 = wszDisplayName; *pch0; ++pch0)
```

Comment:
Is the ':' to be found at a random place in the name? Why not using `wcschr` to find it?

Reply:
The drive icon's caption can contain a colon.
![image](https://user-images.githubusercontent.com/2107452/140602959-ae51bcff-0dc4-454a-9423-127a93442b3f.png)
But a colon is invalid as a filesystem character.

Reply:
Fixed.

Reply:
Invalid characters might be appeared at random positions.

Reply:
if they were invalid it's because they were from past the (initialized) buffer? 🤔

Reply:
No, buffer-overrun is not happened.